### PR TITLE
fix: correct the digest of eip712 when where is a field with empty va…

### LINKED ISSUE #2207: Wrong eip-712 digest hash when empty array occurs in  json
**Describe the bug**
When occurs empty array in json, the `ParamStruct::hashStructJson` will not work well, for example:
```json
{
    "types": {
        "EIP712Domain": [
            {
                "name": "chainId",
                "type": "uint256"
            },
            {
                "name": "name",
                "type": "string"
            },
            {
                "name": "verifyingContract",
                "type": "address"
            },
            {
                "name": "version",
                "type": "string"
            }
        ],
        "Action": [
            {
                "name": "action",
                "type": "string"
            },
            {
                "name": "params",
                "type": "string"
            }
        ],
        "Cell": [
            {
                "name": "capacity",
                "type": "string"
            },
            {
                "name": "lock",
                "type": "string"
            },
            {
                "name": "type",
                "type": "string"
            },
            {
                "name": "data",
                "type": "string"
            },
            {
                "name": "extraData",
                "type": "string"
            }
        ],
        "Transaction": [
            {
                "name": "DAS_MESSAGE",
                "type": "string"
            },
            {
                "name": "inputsCapacity",
                "type": "string"
            },
            {
                "name": "outputsCapacity",
                "type": "string"
            },
            {
                "name": "fee",
                "type": "string"
            },
            {
                "name": "action",
                "type": "Action"
            },
            {
                "name": "inputs",
                "type": "Cell[]"
            },
            {
                "name": "outputs",
                "type": "Cell[]"
            },
            {
                "name": "digest",
                "type": "bytes32"
            }
        ]
    },
    "primaryType": "Transaction",
    "domain": {
        "chainId": 1,
        "name": "da.systems",
        "verifyingContract": "0x0000000000000000000000000000000020210722",
        "version": "1"
    },
    "message": {
        "DAS_MESSAGE": "TRANSFER FROM 0x54366bcd1e73baf55449377bd23123274803236e(906.74221046 CKB) TO ckt1qyqvsej8jggu4hmr45g4h8d9pfkpd0fayfksz44t9q(764.13228446 CKB), 0x54366bcd1e73baf55449377bd23123274803236e(142.609826 CKB)",
        "inputsCapacity": "906.74221046 CKB",
        "outputsCapacity": "906.74211046 CKB",
        "fee": "0.0001 CKB",
        "digest": "0x29cd28dbeb470adb17548563ceb4988953fec7b499e716c16381e5ae4b04021f",
        "action": {
            "action": "transfer",
            "params": "0x00"
        },
        "inputs": [],
        "outputs": []
    }
}
```
The `inputs` and `outputs` is set `[]`, and the outcome is wrong. (Different from MetaMask, which will result in a sig error)

### PR DESCRIPTION
**Pull request by pranksteess**
## Description

When an empty string/array occurs in eip-712 json, the func `ParamStruct::hashStructJson` gives the wrong digest hash. Fixes #2207 

## Testing instructions

I have written the test case in file `eip712_emptyValue.json` and `eip712_emptyArray.json`, the value of some fields are set empty on purpose.

## Types of changes

The expected outcome of the empty string should be returned with the result of keccak256('')

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: src/Ethereum/ABI/Bytes.cpp [hunk 1]
```diff
@@ -125,6 +125,8 @@ Data ParamString::hashStruct() const {
     Data encoded = data(_str);
     if (encoded.size() > 0) {
```
[the remark this answers is not in the document]
Comment (pranksteess) [reply src/Ethereum/ABI/Bytes.cpp]: Fixed

**Comment by pranksteess** [comment]
I found and fixed another bug (#2207), which will result in the same error (wrong digest). Please check it out @catenocrypt

**Comment by pranksteess** [comment]
OK, I rollbacked to the "empty value(string)" version

File: src/Ethereum/ABI/Array.cpp [hunk 2]
```diff
@@ -104,6 +104,9 @@ bool ParamArray::setValueJson(const std::string& value) {
 }
 
 Data ParamArray::hashStruct() const {
+    if (empty) {
```
Comment (catenocrypt) [inline src/Ethereum/ABI/Array.cpp:107]: This is array related, you can leave out this file.

Comment (pranksteess) [reply src/Ethereum/ABI/Array.cpp]: Oh, my bad, missing something

**Review by catenocrypt — APPROVED** [review]

**Pull request merged**